provider/cluster/operatorclients/ip_operator_client.go

```diff
@@ -0,0 +1,195 @@
+package operatorclients
+
+import (
+	"context"
+	"encoding/json"
+	"errors"
+	"fmt"
+	clusterutil "github.com/ovrclk/akash/provider/cluster/util"
+	ipoptypes "github.com/ovrclk/akash/provider/operator/ipoperator/types"
+	mtypes "github.com/ovrclk/akash/x/market/types/v1beta2"
+	"github.com/tendermint/tendermint/libs/log"
+	"io"
+	"k8s.io/client-go/rest"
+	"net"
+	"net/http"
+)
+
+var (
+	errNotImplemented   = errors.New("not implemented")
+	errIPOperatorRemote = errors.New("ip operator remote error")
+)
+
+type IPOperatorClient interface {
+	Check(ctx context.Context) error
+	GetIPAddressUsage(ctx context.Context) (ipoptypes.IPAddressUsage, error)
+
+	GetIPAddressStatus(ctx context.Context, orderID mtypes.OrderID) ([]ipoptypes.LeaseIPStatus, error)
+	Stop()
+	String() string
+}
+
+/* A null client for use in tests and other scenarios */
+type ipOperatorNullClient struct{}
+
+func NullIPOperatorClient() IPOperatorClient {
+	return ipOperatorNullClient{}
+}
+
+func (v ipOperatorNullClient) String() string {
+	return fmt.Sprintf("<%T>", v)
+}
+
+func (ipOperatorNullClient) Check(_ context.Context) error {
+	return errNotImplemented
+}
+
+func (ipOperatorNullClient) GetIPAddressUsage(_ context.Context) (ipoptypes.IPAddressUsage, error) {
+	return ipoptypes.IPAddressUsage{}, errNotImplemented
+}
+
+func (ipOperatorNullClient) Stop() {}
+
+func (ipOperatorNullClient) GetIPAddressStatus(context.Context, mtypes.OrderID) ([]ipoptypes.LeaseIPStatus, error) {
+	return nil, errNotImplemented
+}
+
+func NewIPOperatorClient(logger log.Logger, kubeConfig *rest.Config, endpoint *net.SRV) (IPOperatorClient, error) {
+	sda, err := clusterutil.NewServiceDiscoveryAgent(logger, kubeConfig, "api", "akash-ip-operator", "akash-services", endpoint)
+	if err != nil {
+		return nil, err
+	}
+
+	return &ipOperatorClient{
+		sda: sda,
+		log: logger.With("operator", "ip"),
+	}, nil
+}
+
+func (ipoc *ipOperatorClient) String() string {
+	return fmt.Sprintf("<%T %p>", ipoc, ipoc)
+}
+
+func (ipoc *ipOperatorClient) Stop() {
+	ipoc.sda.Stop()
+}
+
+const (
+	ipOperatorHealthPath = "/health"
+)
+
+/* A client to talk to the Akash implementation of the IP Operator via HTTP */
+type ipOperatorClient struct {
+	sda    clusterutil.ServiceDiscoveryAgent
+	client clusterutil.ServiceClient
+	log    log.Logger
+}
+
+var errNotAlive = errors.New("ip operator is not yet alive")
+
+func (ipoc *ipOperatorClient) Check(ctx context.Context) error {
+	req, err := ipoc.newRequest(ctx, http.MethodGet, ipOperatorHealthPath, nil)
+	if err != nil {
+		return err
+	}
+
+	response, err := ipoc.client.DoRequest(req)
+	if err != nil {
+		return err
+	}
+	ipoc.log.Info("check result", "status", response.StatusCode)
+
+	if response.StatusCode != http.StatusOK {
+		return errNotAlive
+	}
+
+	return nil
+}
+
+func (ipoc *ipOperatorClient) newRequest(ctx context.Context, method string, path string, body io.Reader) (*http.Request, error) {
+	if ipoc.client == nil {
```

Comment:
I think there is a race here around initializing `ipoc.client`, but I can probably just fix it with a mutex.